### PR TITLE
fix(telemetry): Change sensor names for FlySky GPS Lat and Lon sensors.

### LINKED ISSUE #4062: GPS telemetry  variable names (lat/lon both named "GPS")
### Is there an existing issue for this problem?

- [X] I have searched the existing issues

### What part of EdgeTX is the focus of this bug?

Other (Please specify below)

### Current Behavior

Using OxS on RP2040 and connecting a GPS to it, I got a working module to send telemetry via my FS-IA6B receiver FlSky2A protocol.

Telemetry is discovered but latitude and longitude have the same "GPS" name. (which is not useful in lua scripts).

Mstrens (author of OxS on RP2040) told me this :

"_In flysky (Ibus) protocol, oXs does not send the name of the field but a code.
The codes are defined like this:
IBUS_SENSOR_TYPE_GPS_LAT = 0x80, //4bytes signed WGS84 in degrees * 1E7
IBUS_SENSOR_TYPE_GPS_LON = 0x81, //4bytes signed WGS84 in degrees * 1E7

So oXs uses 2 different codes for long and lat.
I expect that it is EdgeTX that converts the code into a free text.
I suggest that you ask on EdgeTx forum because I expect the issue is at this level._"

For clarity here is the picture showing the "bug"

name1

I had to change the names:
name2

This is of course not a critical bug. But every time I "re discover" telemetry the names are reset to "GPS" !

### Expected Behavior

change "GPS" by "Lat" and "Lon" (or "Glat" "Glon")

Thanks

### Steps To Reproduce

discover a telemetry with GPS inside.

### Version

2.9.0

### Transmitter

Radiomaster TX16S / TX16SMK2

### Operating System (OS)

_No response_

### OS Version

_No response_

### Anything else?

full explaination of the project here : https://hackaday.io/project/192752-variometer-gps-tracker-remoteid-for-rc-glider/details

(bug explained after the middle of the details page)

image

### PR DESCRIPTION
**Pull request by philmoz**
Fixes #4062 

Add "GLat" and "GLon" names for FlySky GPS telemetry sensors.

### CODE REVIEW

**Comment by pfeerick** [comment]
More is probably needed for consistency... ie. Crossfire and Ghost are the same

**Review by frankiearzu — COMMENTED** [review]
In general, i think we should use a single GPS sensor with the Lat/Long combined.. This format is expected to properly display it in the screen, as well as there are some sensor operations to compute distance from a fix coordinate (probably distance from home).

Was this code tested with the sensors???    Im in the process of getting the components for buiding a 0xS.. only have Spektrum sensors.

1. i think the changes for crossfire.cpp will not work... is using the same sensor ID (GPS_ID+0).. so you will not have Lat and Long separate.... I think this one probably worked correctly before by populating a GPS with latitude/longitude combined into a single sensor.

2. flysky_ibus.cpp:   This will split Lat/Long into two independent sensors, but for consistency and usage of GPS coordinates in other parts of EdgeTx, i think is better to have them combined into a single sensor.. The code corectly split them, but is missing a value conversion (probably will need a division by 10 or 100).   See my comment in the ticket.. setup the code to do it, but don't have a sensor to test.

3. ghost.cpp will probably need similar procesing as flysky_ibus.cpp

**Comment by 3djc** [comment]
Splitting sensors puts even more pressure on 128 radio where sensors count is even lower

**Comment by philmoz** [comment]
This change does not split GPS into two sensors - this is what is already being done in the code for the data coming in from the telemetry stream for the affected protocols. It merely renames the two existing sensors to differentiate them.

**Comment by frankiearzu** [comment]
@philmoz 
Not trying to be an ass.. just trying to do what is right.  Looked at more telemetry code, seems that single GPS sensor is the standard for many of them (crossfire,ghost, spektrum)

I think the important question is, do we want to standarized in a single GPS sensor with Lat/Long togeher or keep Lat/Log separate??   From what i have seen in the EdgeTX code. there are many places that needs it to be a single sensor.

Second question?? was the code tested with the a sensor with the 3 different protocols??  I think the crossfire code and ghost was working as expected... by renaming the sensor, we will break it.  i.e.  We will end up having only the "GLAT" sensor with both Lat and Long together. The "GLON" might not even show up.

As for the the code changes  (based on the original code).
1. crosfire.cpp code seems that it was using a single sensor, and i think was working correctly.  As you can see the ID of the sensor will be (GPS_ID+0) for both Latitude and Longitude..  even when they are in separate rows, is a single sensor.   This code is correctly populating the LAT portion of the sensor (UNIT_GPS_LATITUDE) as well as the Longitude (UNIT_GPS_LONGITUDE).   If you change the names on both rows as done in your code, it will only use the first one... so you will have a GLAT sensor with both Latitude/longitude, and no longitude.

image

2. FrSky_ibus.cpp:   The code can process sensors in two ways,  individual Lat/Log packets, or a packet with all of them combined.    To create a single sensor is simple:
a) create a single virtual sensor id representing both  (or we can just use LAT as the representative for both)
image
b)  We have to add special processing for the individual sensors to populate the single virtual sensor:
image

3. ghost.cpp:  Seems that it was working before by populating a single GPS value.. you can see in the code that when the LONG is processed. the code redirected it to the LAT sensor, but depending of the sensor.unit, it will populate the Lat or Long portion of a single sensor (LAT represents both).   The only question i have in this code is if UNIT_GPS_LATITUDE will be treated correctly for display (UNIT_GPS).. otherwise we have to create a fake/virtual sensor with the right unit type.. not sure until we test with a real sensor.

image

In summary,
Looks like frysky_ibus was the only one not working, and by changing the name, the new code is breaking crossfire and ghost protocols.   Thats why my 2nd question... was this new code tested with all 3 protocols???

As you can see, single sensor GPS seems to be the standard.  the bug was only on the frsky_ibus, but is easilly fixed to be a single sensor.

**Comment by pfeerick** [comment]
Also need to consider that there is flysky_ibus AND flysky_nv14 - which haven't been kept in sync, and IIRC one of them really needs to be done away with, but that is for another discussion. 

As long as there is sufficient handling for the fact that the GPS sensor has two separate values associated with it (i.e. so widgets show both values  - which they used to for crossfire, and logical switches - no sure what happens there) ... if the majority of the telemetry processing so far is to treat it as a single sensor, you're right, it should be the other way. 

I can probably test the Ghost scenario at some point as I have Ghost and an iNav board I can hook a GPS up to.

**Comment by frankiearzu** [comment]
Review the code of Ghost.. and i ansered my own question.. at the end the sensor if of type UNIT_GPS that is correct.  I think everything is correct.   The only think that i don't like is that we have 3 "hacks" in different places of the ghost.cpp to properly handle GPS, but should works. 

image
